netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -365,10 +437,22 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        * lifetime and we only want to configure it once.
        */
       if (handlers != null) {
-        ctx.pipeline().addFirst(handlers);
+        for (ChannelHandler handler : handlers) {
+          ctx.pipeline().addBefore(ctx.name(), null, handler);
+        }
+        ChannelHandler handler0 = handlers[0];
```

Comment:
The channelRegistered event is delivered to the front. By adding these handlers before the current one, they won't receive the event. But the nested negotiator needs the channelRegistered event. Thus I trigger the event to the handlers that missed it. That causes this handler to receive the event twice, so I squelch one of the two.

It would be possible to teach the negotiation handlers to observe other events, but it would be rare that that code would be run, so it seems more bug-prone.